{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from eda_imports import *\n",
    "\n",
    "import pysam\n",
    "\n",
    "import kleat.misc.settings as S\n",
    "import kleat.misc.utils as U"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "lele = pd.read_pickle('../benchmark_transcriptome/after-num-suffix-reads-fix-renamed-header/HBRC4.ml_ready.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(931053, 66)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lele.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>strand</th>\n",
       "      <th>clv</th>\n",
       "      <th>aclv</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>signed_dist_to_aclv</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>contig_ids_at_pos</th>\n",
       "      <th>any_contig_is_hardclipped</th>\n",
       "      <th>contig_max_len</th>\n",
       "      <th>contig_max_mapq</th>\n",
       "      <th>num_contigs_suffix</th>\n",
       "      <th>num_contigs_bridge</th>\n",
       "      <th>num_contigs_link</th>\n",
       "      <th>num_contigs_blank</th>\n",
       "      <th>num_total_contigs</th>\n",
       "      <th>num_reads_suffix</th>\n",
       "      <th>num_reads_bridge</th>\n",
       "      <th>num_reads_link</th>\n",
       "      <th>max_read_tail_len_suffix</th>\n",
       "      <th>max_read_tail_len_bridge</th>\n",
       "      <th>max_contig_tail_len_suffix</th>\n",
       "      <th>ctg_hex</th>\n",
       "      <th>ctg_hex_id</th>\n",
       "      <th>ctg_hex_pos</th>\n",
       "      <th>ctg_hex_dist</th>\n",
       "      <th>ref_hex</th>\n",
       "      <th>ref_hex_id</th>\n",
       "      <th>ref_hex_pos</th>\n",
       "      <th>ref_hex_dist</th>\n",
       "      <th>abs_dist_to_aclv</th>\n",
       "      <th>ctg_AAAAAG</th>\n",
       "      <th>ctg_AAAACA</th>\n",
       "      <th>ctg_AAGAAA</th>\n",
       "      <th>ctg_AATAAA</th>\n",
       "      <th>ctg_AATACA</th>\n",
       "      <th>ctg_AATAGA</th>\n",
       "      <th>ctg_AATATA</th>\n",
       "      <th>ctg_AATGAA</th>\n",
       "      <th>ctg_ACTAAA</th>\n",
       "      <th>ctg_AGTAAA</th>\n",
       "      <th>ctg_ATTAAA</th>\n",
       "      <th>ctg_CATAAA</th>\n",
       "      <th>ctg_GATAAA</th>\n",
       "      <th>ctg_GGGGCT</th>\n",
       "      <th>ctg_NA</th>\n",
       "      <th>ctg_TATAAA</th>\n",
       "      <th>ctg_TTTAAA</th>\n",
       "      <th>ref_AAAAAG</th>\n",
       "      <th>ref_AAAACA</th>\n",
       "      <th>ref_AAGAAA</th>\n",
       "      <th>ref_AATAAA</th>\n",
       "      <th>ref_AATACA</th>\n",
       "      <th>ref_AATAGA</th>\n",
       "      <th>ref_AATATA</th>\n",
       "      <th>ref_AATGAA</th>\n",
       "      <th>ref_ACTAAA</th>\n",
       "      <th>ref_AGTAAA</th>\n",
       "      <th>ref_ATTAAA</th>\n",
       "      <th>ref_CATAAA</th>\n",
       "      <th>ref_GATAAA</th>\n",
       "      <th>ref_GGGGCT</th>\n",
       "      <th>ref_NA</th>\n",
       "      <th>ref_TATAAA</th>\n",
       "      <th>ref_TTTAAA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>12556</td>\n",
       "      <td>13671</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>-1115</td>\n",
       "      <td>blank</td>\n",
       "      <td>01.S.508265@110</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAGAAA</td>\n",
       "      <td>5</td>\n",
       "      <td>12543</td>\n",
       "      <td>13</td>\n",
       "      <td>AAGAAA</td>\n",
       "      <td>5</td>\n",
       "      <td>12543</td>\n",
       "      <td>13</td>\n",
       "      <td>1115</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>14602</td>\n",
       "      <td>14413</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>189</td>\n",
       "      <td>blank</td>\n",
       "      <td>E2.L.2150@240</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>189</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname strand    clv   aclv gene_name          gene_id  signed_dist_to_aclv evidence_type contig_ids_at_pos any_contig_is_hardclipped contig_max_len contig_max_mapq num_contigs_suffix num_contigs_bridge num_contigs_link num_contigs_blank num_total_contigs num_reads_suffix num_reads_bridge  \\\n",
       "0  chr1    +      12556  13671  DDX11L1   ENSG00000223972 -1115                 blank         01.S.508265@110   True                      3              0               0                  0                  0                1                 1                 0                0                 \n",
       "1  chr1    +      14602  14413  DDX11L1   ENSG00000223972  189                  blank         E2.L.2150@240     True                      3              0               0                  0                  0                1                 1                 0                0                 \n",
       "\n",
       "  num_reads_link max_read_tail_len_suffix max_read_tail_len_bridge max_contig_tail_len_suffix ctg_hex ctg_hex_id ctg_hex_pos ctg_hex_dist ref_hex ref_hex_id ref_hex_pos ref_hex_dist  abs_dist_to_aclv  ctg_AAAAAG  ctg_AAAACA  ctg_AAGAAA  ctg_AATAAA  ctg_AATACA  ctg_AATAGA  ctg_AATATA  ctg_AATGAA  \\\n",
       "0  0              0                        0                        0                          AAGAAA  5          12543       13           AAGAAA  5          12543       13           1115              0           0           1           0           0           0           0           0            \n",
       "1  0              0                        0                        0                          NA      -1         -1          -1           NA      -1         -1          -1           189               0           0           0           0           0           0           0           0            \n",
       "\n",
       "   ctg_ACTAAA  ctg_AGTAAA  ctg_ATTAAA  ctg_CATAAA  ctg_GATAAA  ctg_GGGGCT  ctg_NA  ctg_TATAAA  ctg_TTTAAA  ref_AAAAAG  ref_AAAACA  ref_AAGAAA  ref_AATAAA  ref_AATACA  ref_AATAGA  ref_AATATA  ref_AATGAA  ref_ACTAAA  ref_AGTAAA  ref_ATTAAA  ref_CATAAA  ref_GATAAA  ref_GGGGCT  ref_NA  ref_TATAAA  \\\n",
       "0  0           0           0           0           0           0           0       0           0           0           0           1           0           0           0           0           0           0           0           0           0           0           0           0       0            \n",
       "1  0           0           0           0           0           0           1       0           0           0           0           0           0           0           0           0           0           0           0           0           0           0           0           1       0            \n",
       "\n",
       "   ref_TTTAAA  \n",
       "0  0           \n",
       "1  0           "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lele.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = []\n",
    "for i in glob.glob('../max_recall.*'):\n",
    "    dfs.append(pd.read_csv(i, sep='\\t'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ndf = pd.concat(dfs).sort_values('sample_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBRC4</td>\n",
       "      <td>0.026774</td>\n",
       "      <td>0.242732</td>\n",
       "      <td>0.048228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBRC6</td>\n",
       "      <td>0.014323</td>\n",
       "      <td>0.359445</td>\n",
       "      <td>0.027547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UHRC1</td>\n",
       "      <td>0.025576</td>\n",
       "      <td>0.198304</td>\n",
       "      <td>0.045308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UHRC2</td>\n",
       "      <td>0.023578</td>\n",
       "      <td>0.222404</td>\n",
       "      <td>0.042635</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id  precision    recall        f1\n",
       "0  HBRC4     0.026774   0.242732  0.048228\n",
       "0  HBRC6     0.014323   0.359445  0.027547\n",
       "0  UHRC1     0.025576   0.198304  0.045308\n",
       "0  UHRC2     0.023578   0.222404  0.042635"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_kleat2 = pd.read_csv('/projects/btl/zxue/tasrkleat-TCGA-results/tasrkleat-TCGA-analysis-scripts/benchmark-kleat.bk/UHR/C1/tasrkleat-results/kleat/cba.KLEAT', sep='\\t')\n",
    "df_kleat2.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>transcript_strand</th>\n",
       "      <th>coding</th>\n",
       "      <th>contig</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>cleavage_site</th>\n",
       "      <th>within_UTR</th>\n",
       "      <th>distance_from_annotated_site</th>\n",
       "      <th>ESTs</th>\n",
       "      <th>length_of_tail_in_contig</th>\n",
       "      <th>number_of_tail_reads</th>\n",
       "      <th>number_of_bridge_reads</th>\n",
       "      <th>max_bridge_read_tail_length</th>\n",
       "      <th>bridge_read_identities</th>\n",
       "      <th>tail+bridge_reads</th>\n",
       "      <th>number_of_link_pairs</th>\n",
       "      <th>max_link_pair_length</th>\n",
       "      <th>link_pair_identities</th>\n",
       "      <th>hexamer_loc+id</th>\n",
       "      <th>3UTR_start_end</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2058</th>\n",
       "      <td>GNAS</td>\n",
       "      <td>ENST00000371095</td>\n",
       "      <td>+</td>\n",
       "      <td>yes</td>\n",
       "      <td>A0.R100623,A0.J100905</td>\n",
       "      <td>chr20</td>\n",
       "      <td>57486247</td>\n",
       "      <td>yes</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>FGFR3</td>\n",
       "      <td>ENST00000481110</td>\n",
       "      <td>+</td>\n",
       "      <td>yes</td>\n",
       "      <td>A0.R100856,A0.J100703</td>\n",
       "      <td>chr4</td>\n",
       "      <td>1810594</td>\n",
       "      <td>yes</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>1810573:1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1817</th>\n",
       "      <td>SMARCA4</td>\n",
       "      <td>ENST00000450717</td>\n",
       "      <td>+</td>\n",
       "      <td>yes</td>\n",
       "      <td>A0.R100671,A1.S26663,A0.J100737</td>\n",
       "      <td>chr19</td>\n",
       "      <td>11172953</td>\n",
       "      <td>yes</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>11172927:1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1284</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>ENST00000407796</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>A1.R26107,A1.S26628,A0.R100669</td>\n",
       "      <td>chr14</td>\n",
       "      <td>105235686</td>\n",
       "      <td>yes</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>105235705:1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>RHOBTB3</td>\n",
       "      <td>ENST00000379982</td>\n",
       "      <td>+</td>\n",
       "      <td>yes</td>\n",
       "      <td>A0.R100765</td>\n",
       "      <td>chr5</td>\n",
       "      <td>95130822</td>\n",
       "      <td>yes</td>\n",
       "      <td>1249</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1040</th>\n",
       "      <td>HRAS</td>\n",
       "      <td>ENST00000451590</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>A1.R26175,A1.R26487,A1.S26555</td>\n",
       "      <td>chr11</td>\n",
       "      <td>532242</td>\n",
       "      <td>yes</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>532264:3</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1880</th>\n",
       "      <td>CEBPA</td>\n",
       "      <td>ENST00000498907</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>A0.S101662</td>\n",
       "      <td>chr19</td>\n",
       "      <td>33790840</td>\n",
       "      <td>yes</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>33790859:4</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>HNRNPA2B1</td>\n",
       "      <td>ENST00000356674</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>A0.S32990</td>\n",
       "      <td>chr7</td>\n",
       "      <td>26232870</td>\n",
       "      <td>yes</td>\n",
       "      <td>1407</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1396</th>\n",
       "      <td>BLOC1S6</td>\n",
       "      <td>ENST00000567523</td>\n",
       "      <td>+</td>\n",
       "      <td>yes</td>\n",
       "      <td>A0.R101303</td>\n",
       "      <td>chr15</td>\n",
       "      <td>45899979</td>\n",
       "      <td>yes</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1547</th>\n",
       "      <td>CDH1</td>\n",
       "      <td>ENST00000261769</td>\n",
       "      <td>+</td>\n",
       "      <td>yes</td>\n",
       "      <td>A0.R100743</td>\n",
       "      <td>chr16</td>\n",
       "      <td>68869444</td>\n",
       "      <td>yes</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene       transcript transcript_strand coding                           contig chromosome  cleavage_site within_UTR  distance_from_annotated_site ESTs  length_of_tail_in_contig  number_of_tail_reads  number_of_bridge_reads  max_bridge_read_tail_length bridge_read_identities  \\\n",
       "2058  GNAS       ENST00000371095  +                 yes    A0.R100623,A0.J100905            chr20      57486247       yes        0                             0    16                        20                    0                       0                            -                       \n",
       "499   FGFR3      ENST00000481110  +                 yes    A0.R100856,A0.J100703            chr4       1810594        yes        2                             0    13                        12                    0                       0                            -                       \n",
       "1817  SMARCA4    ENST00000450717  +                 yes    A0.R100671,A1.S26663,A0.J100737  chr19      11172953       yes        5                             0    14                        9                     0                       0                            -                       \n",
       "1284  AKT1       ENST00000407796  -                 yes    A1.R26107,A1.S26628,A0.R100669   chr14      105235686      yes        1                             0    12                        8                     0                       0                            -                       \n",
       "604   RHOBTB3    ENST00000379982  +                 yes    A0.R100765                       chr5       95130822       yes        1249                          0    4                         8                     0                       0                            -                       \n",
       "1040  HRAS       ENST00000451590  -                 yes    A1.R26175,A1.R26487,A1.S26555    chr11      532242         yes        0                             0    11                        7                     0                       0                            -                       \n",
       "1880  CEBPA      ENST00000498907  -                 yes    A0.S101662                       chr19      33790840       yes        1                             0    9                         7                     0                       0                            -                       \n",
       "724   HNRNPA2B1  ENST00000356674  -                 yes    A0.S32990                        chr7       26232870       yes        1407                          0    1                         6                     0                       0                            -                       \n",
       "1396  BLOC1S6    ENST00000567523  +                 yes    A0.R101303                       chr15      45899979       yes        8                             0    2                         6                     0                       0                            -                       \n",
       "1547  CDH1       ENST00000261769  +                 yes    A0.R100743                       chr16      68869444       yes        7                             0    8                         6                     0                       0                            -                       \n",
       "\n",
       "      tail+bridge_reads  number_of_link_pairs  max_link_pair_length link_pair_identities hexamer_loc+id 3UTR_start_end  flag  \n",
       "2058  20                 0                     0                     -                    -              -              1     \n",
       "499   12                 0                     0                     -                    1810573:1      -              1     \n",
       "1817  9                  0                     0                     -                    11172927:1     -              1     \n",
       "1284  8                  0                     0                     -                    105235705:1    -              1     \n",
       "604   8                  0                     0                     -                    -              -              1     \n",
       "1040  7                  0                     0                     -                    532264:3       -              1     \n",
       "1880  7                  0                     0                     -                    33790859:4     -              1     \n",
       "724   6                  0                     0                     -                    -              -              1     \n",
       "1396  6                  0                     0                     -                    -              -              1     \n",
       "1547  6                  0                     0                     -                    -              -              1     "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kleat2.sort_values('number_of_tail_reads', ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.insert(0, '../../kleat/benchmark_scripts/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from train_arbor import cluster_clv, compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from max_recall_analysis import cluster_clv_parallel, map_to_ref"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare current version and a previous ref-hex-dist version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "cur_df = pd.read_pickle('./../../kleat/benchmark_transcriptome/current-buggy/UHRC1.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(869893, 30)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cur_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2018-07-17 14:29:05,229|INFO|grouping by ['seqname', 'strand']\n",
      "100%|██████████| 50/50 [00:00<00:00, 139.83it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clustering clvs in parallel) using 24 CPUs ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2018-07-17 14:30:00,349|INFO|concatenating clustered sub dataframes ...\n",
      "2018-07-17 14:30:01,181|INFO|dropping duplicates after clustering ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 21.6 s, sys: 3.5 s, total: 25.1 s\n",
      "Wall time: 57 s\n"
     ]
    }
   ],
   "source": [
    "%time cur_df_clustered = cluster_clv_parallel(cur_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.4 s, sys: 460 ms, total: 2.86 s\n",
      "Wall time: 2.86 s\n"
     ]
    }
   ],
   "source": [
    "%time ref_df = pd.read_pickle('./../../kleat/benchmark_transcriptome/after_chimeric-ref-hex/UHRC1.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(869986, 30)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2018-07-17 14:30:38,580|INFO|grouping by ['seqname', 'strand']\n",
      "100%|██████████| 50/50 [00:00<00:00, 163.00it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clustering clvs in parallel) using 24 CPUs ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2018-07-17 14:31:33,376|INFO|concatenating clustered sub dataframes ...\n",
      "2018-07-17 14:31:34,184|INFO|dropping duplicates after clustering ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 21.9 s, sys: 3.01 s, total: 24.9 s\n",
      "Wall time: 56.9 s\n"
     ]
    }
   ],
   "source": [
    "%time ref_df_clustered = cluster_clv_parallel(ref_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(805439, 3)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df_clustered.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2018-07-17 14:33:38,391|INFO|mapping predicted clv to ground truth ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mapping ...\n"
     ]
    }
   ],
   "source": [
    "print('mapping ...')\n",
    "map_cutoff = 50\n",
    "df_mapped = map_to_ref(cur_df_clustered, ref_df_clustered, map_cutoff)\n",
    "df_mapped['is_tp'] = df_mapped.abs_dist < map_cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "calculating metrics ...\n"
     ]
    }
   ],
   "source": [
    "print('calculating metrics ...')\n",
    "recall = df_mapped.query('is_tp').shape[0] / ref_df_clustered.shape[0]\n",
    "precision = df_mapped.query('is_tp').shape[0] / df_mapped.shape[0]\n",
    "f1 = (2 * recall * precision) / (recall + precision)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9997479250381217\t0.9995952517819475\t0.9996715825808393\n"
     ]
    }
   ],
   "source": [
    "print(f'{precision}\\t{recall}\\t{f1}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>strand</th>\n",
       "      <th>clv</th>\n",
       "      <th>mapped_ref_clv</th>\n",
       "      <th>dist</th>\n",
       "      <th>abs_dist</th>\n",
       "      <th>is_tp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>12199</td>\n",
       "      <td>12199</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>14748</td>\n",
       "      <td>14748</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname strand    clv  mapped_ref_clv  dist  abs_dist  is_tp\n",
       "0  chr1    +      12199  12199           0     0         True \n",
       "1  chr1    +      14748  14748           0     0         True "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mapped.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cur_df['clv_m1'] = cur_df.clv - 1\n",
    "cur_df['clv_p1'] = cur_df.clv + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "cur_clvs_m1 = set([tuple(_) for _ in cur_df[['seqname', 'strand', 'clv_m1']].values])\n",
    "cur_clvs    = set([tuple(_) for _ in cur_df[['seqname', 'strand', 'clv'   ]].values])\n",
    "cur_clvs_p1 = set([tuple(_) for _ in cur_df[['seqname', 'strand', 'clv_p1']].values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "cur_clvs_set = cur_clvs_m1 | cur_clvs | cur_clvs_p1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2577194"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cur_clvs_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2609679"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cur_clvs_m1) + len(cur_clvs) + len(cur_clvs_p1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.35 s, sys: 32.9 ms, total: 1.38 s\n",
      "Wall time: 1.38 s\n"
     ]
    }
   ],
   "source": [
    "%time ref_clvs = set([tuple(_) for _ in ref_df[['seqname', 'strand', 'clv']].values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "534"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ref_clvs - cur_clvs_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1707742"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cur_clvs_set - ref_clvs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "missed = ref_clvs - cur_clvs_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('chr7', '+', 73660993),\n",
       " ('chr6', '+', 75950892),\n",
       " ('chr19', '+', 649680),\n",
       " ('chr16', '-', 78534894),\n",
       " ('chr13', '+', 40914558)]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(missed)[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "missed_clvs = [_[2] for _ in missed]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "missed_df = ref_df.query(f'clv in {missed_clvs}').copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "bridge    534\n",
       "blank     10 \n",
       "Name: evidence_type, dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missed_df.evidence_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "missed_df['abs_dist_to_aclv'] = missed_df.signed_dist_to_aclv.abs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>strand</th>\n",
       "      <th>clv</th>\n",
       "      <th>aclv</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>signed_dist_to_aclv</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>contig_ids</th>\n",
       "      <th>max_contig_len</th>\n",
       "      <th>max_contig_mapq</th>\n",
       "      <th>any_contig_is_hardclipped</th>\n",
       "      <th>num_suffix_contigs</th>\n",
       "      <th>num_bridge_contigs</th>\n",
       "      <th>num_link_contigs</th>\n",
       "      <th>num_blank_contigs</th>\n",
       "      <th>num_total_contigs</th>\n",
       "      <th>num_suffix_reads</th>\n",
       "      <th>max_suffix_contig_tail_len</th>\n",
       "      <th>num_bridge_reads</th>\n",
       "      <th>max_bridge_read_tail_len</th>\n",
       "      <th>num_link_reads</th>\n",
       "      <th>ctg_hex</th>\n",
       "      <th>ctg_hex_id</th>\n",
       "      <th>ctg_hex_pos</th>\n",
       "      <th>ctg_hex_dist</th>\n",
       "      <th>ref_hex</th>\n",
       "      <th>ref_hex_id</th>\n",
       "      <th>ref_hex_pos</th>\n",
       "      <th>ref_hex_dist</th>\n",
       "      <th>abs_dist_to_aclv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>657410</th>\n",
       "      <td>chr5</td>\n",
       "      <td>-</td>\n",
       "      <td>74017080</td>\n",
       "      <td>74017083</td>\n",
       "      <td>GFM2</td>\n",
       "      <td>ENSG00000164347</td>\n",
       "      <td>-3</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E0.L.64191</td>\n",
       "      <td>963</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>74017103</td>\n",
       "      <td>23</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>74017105</td>\n",
       "      <td>25</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798032</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>144668458</td>\n",
       "      <td>144668388</td>\n",
       "      <td>EEF1D</td>\n",
       "      <td>ENSG00000104529</td>\n",
       "      <td>70</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E1.L.11467</td>\n",
       "      <td>1298</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>144668497</td>\n",
       "      <td>39</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>144668497</td>\n",
       "      <td>39</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824660</th>\n",
       "      <td>chr9</td>\n",
       "      <td>-</td>\n",
       "      <td>99314154</td>\n",
       "      <td>99314082</td>\n",
       "      <td>CDC14B</td>\n",
       "      <td>ENSG00000081377</td>\n",
       "      <td>72</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E1.L.4856</td>\n",
       "      <td>2381</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>99314188</td>\n",
       "      <td>34</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>99314188</td>\n",
       "      <td>34</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245655</th>\n",
       "      <td>chr14</td>\n",
       "      <td>-</td>\n",
       "      <td>50050393</td>\n",
       "      <td>50050296</td>\n",
       "      <td>RPS29</td>\n",
       "      <td>ENSG00000213741</td>\n",
       "      <td>97</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E1.L.7611</td>\n",
       "      <td>401</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>50050420</td>\n",
       "      <td>27</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>50050420</td>\n",
       "      <td>27</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439840</th>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>170492976</td>\n",
       "      <td>170492752</td>\n",
       "      <td>PPIG</td>\n",
       "      <td>ENSG00000138398</td>\n",
       "      <td>224</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E0.L.63481</td>\n",
       "      <td>374</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>170492937</td>\n",
       "      <td>39</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443718</th>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>198367763</td>\n",
       "      <td>198368020</td>\n",
       "      <td>HSPE1</td>\n",
       "      <td>ENSG00000115541</td>\n",
       "      <td>-257</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E3.L.197</td>\n",
       "      <td>616</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>198367738</td>\n",
       "      <td>25</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>198367738</td>\n",
       "      <td>25</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477802</th>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>198360099</td>\n",
       "      <td>198359774</td>\n",
       "      <td>HSPD1</td>\n",
       "      <td>ENSG00000144381</td>\n",
       "      <td>325</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E3.L.96</td>\n",
       "      <td>1923</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>198360134</td>\n",
       "      <td>35</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>198360134</td>\n",
       "      <td>35</td>\n",
       "      <td>325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>833396</th>\n",
       "      <td>chrM</td>\n",
       "      <td>+</td>\n",
       "      <td>997</td>\n",
       "      <td>648</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>349</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E0.L.43441</td>\n",
       "      <td>274</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>967</td>\n",
       "      <td>30</td>\n",
       "      <td>349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550652</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>101284073</td>\n",
       "      <td>101284562</td>\n",
       "      <td>TRMT10C</td>\n",
       "      <td>ENSG00000174173</td>\n",
       "      <td>-489</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E0.S.109816</td>\n",
       "      <td>106</td>\n",
       "      <td>40</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>101284065</td>\n",
       "      <td>8</td>\n",
       "      <td>489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712114</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>135307948</td>\n",
       "      <td>135308791</td>\n",
       "      <td>HBS1L</td>\n",
       "      <td>ENSG00000112339</td>\n",
       "      <td>-843</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E2.L.2538</td>\n",
       "      <td>2699</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>135307968</td>\n",
       "      <td>20</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>135307968</td>\n",
       "      <td>20</td>\n",
       "      <td>843</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname strand        clv       aclv gene_name          gene_id  signed_dist_to_aclv evidence_type   contig_ids max_contig_len max_contig_mapq any_contig_is_hardclipped num_suffix_contigs num_bridge_contigs num_link_contigs num_blank_contigs num_total_contigs num_suffix_reads  \\\n",
       "657410  chr5    -      74017080   74017083   GFM2      ENSG00000164347 -3                    bridge        E0.L.64191   963            40              False                     0                  1                  0                0                 1                 0                 \n",
       "798032  chr8    -      144668458  144668388  EEF1D     ENSG00000104529  70                   bridge        E1.L.11467   1298           40              False                     0                  1                  0                0                 1                 0                 \n",
       "824660  chr9    -      99314154   99314082   CDC14B    ENSG00000081377  72                   bridge        E1.L.4856    2381           40              False                     0                  1                  0                0                 1                 0                 \n",
       "245655  chr14   -      50050393   50050296   RPS29     ENSG00000213741  97                   bridge        E1.L.7611    401            40              False                     0                  1                  0                0                 1                 0                 \n",
       "439840  chr2    +      170492976  170492752  PPIG      ENSG00000138398  224                  bridge        E0.L.63481   374            0               True                      0                  1                  0                0                 1                 0                 \n",
       "443718  chr2    +      198367763  198368020  HSPE1     ENSG00000115541 -257                  bridge        E3.L.197     616            40              False                     0                  1                  0                0                 1                 0                 \n",
       "477802  chr2    -      198360099  198359774  HSPD1     ENSG00000144381  325                  bridge        E3.L.96      1923           40              False                     0                  1                  0                0                 1                 0                 \n",
       "833396  chrM    +      997        648        MT-TF     ENSG00000210049  349                  bridge        E0.L.43441   274            0               True                      0                  1                  0                0                 1                 0                 \n",
       "550652  chr3    +      101284073  101284562  TRMT10C   ENSG00000174173 -489                  bridge        E0.S.109816  106            40              True                      0                  1                  0                0                 1                 0                 \n",
       "712114  chr6    -      135307948  135308791  HBS1L     ENSG00000112339 -843                  bridge        E2.L.2538    2699           40              False                     0                  1                  0                0                 1                 0                 \n",
       "\n",
       "       max_suffix_contig_tail_len num_bridge_reads max_bridge_read_tail_len num_link_reads ctg_hex ctg_hex_id ctg_hex_pos ctg_hex_dist ref_hex ref_hex_id ref_hex_pos ref_hex_dist  abs_dist_to_aclv  \n",
       "657410  0                          2                5                        0              AATAAA  16         74017103    23           AATAAA  16         74017105    25           3                 \n",
       "798032  0                          1                1                        0              AATAAA  16         144668497   39           AATAAA  16         144668497   39           70                \n",
       "824660  0                          2                2                        0              AATAAA  16         99314188    34           AATAAA  16         99314188    34           72                \n",
       "245655  0                          2                2                        0              AATAAA  16         50050420    27           AATAAA  16         50050420    27           97                \n",
       "439840  0                          2                1                        0              NA      -1         -1          -1           AATAAA  16         170492937   39           224               \n",
       "443718  0                          4                2                        0              AATAAA  16         198367738   25           AATAAA  16         198367738   25           257               \n",
       "477802  0                          6                2                        0              AATAAA  16         198360134   35           AATAAA  16         198360134   35           325               \n",
       "833396  0                          1                2                        0              NA      -1         -1          -1           AATAAA  16         967         30           349               \n",
       "550652  0                          6                5                        0              NA      -1         -1          -1           AATAAA  16         101284065   8            489               \n",
       "712114  0                          2                2                        0              AATAAA  16         135307968   20           AATAAA  16         135307968   20           843               "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missed_df.query('evidence_type == \"bridge\"')\\\n",
    "    .query('ref_hex == \"AATAAA\"').sort_values('abs_dist_to_aclv').head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>strand</th>\n",
       "      <th>clv</th>\n",
       "      <th>aclv</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>signed_dist_to_aclv</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>contig_ids</th>\n",
       "      <th>max_contig_len</th>\n",
       "      <th>max_contig_mapq</th>\n",
       "      <th>any_contig_is_hardclipped</th>\n",
       "      <th>num_suffix_contigs</th>\n",
       "      <th>num_bridge_contigs</th>\n",
       "      <th>num_link_contigs</th>\n",
       "      <th>num_blank_contigs</th>\n",
       "      <th>num_total_contigs</th>\n",
       "      <th>num_suffix_reads</th>\n",
       "      <th>max_suffix_contig_tail_len</th>\n",
       "      <th>num_bridge_reads</th>\n",
       "      <th>max_bridge_read_tail_len</th>\n",
       "      <th>num_link_reads</th>\n",
       "      <th>ctg_hex</th>\n",
       "      <th>ctg_hex_id</th>\n",
       "      <th>ctg_hex_pos</th>\n",
       "      <th>ctg_hex_dist</th>\n",
       "      <th>ref_hex</th>\n",
       "      <th>ref_hex_id</th>\n",
       "      <th>ref_hex_pos</th>\n",
       "      <th>ref_hex_dist</th>\n",
       "      <th>clv_m1</th>\n",
       "      <th>clv_p1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>781363</th>\n",
       "      <td>chr8</td>\n",
       "      <td>+</td>\n",
       "      <td>144677134</td>\n",
       "      <td>144682485</td>\n",
       "      <td>TIGD5</td>\n",
       "      <td>ENSG00000179886</td>\n",
       "      <td>-5351</td>\n",
       "      <td>blank</td>\n",
       "      <td>E1.L.11467</td>\n",
       "      <td>1298</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>144677133</td>\n",
       "      <td>144677135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797962</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>144663470</td>\n",
       "      <td>144663248</td>\n",
       "      <td>EEF1D</td>\n",
       "      <td>ENSG00000104529</td>\n",
       "      <td>222</td>\n",
       "      <td>blank</td>\n",
       "      <td>E0.S.110148|E1.L.11467</td>\n",
       "      <td>1298</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>144663469</td>\n",
       "      <td>144663471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797971</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>144668898</td>\n",
       "      <td>144668898</td>\n",
       "      <td>EEF1D</td>\n",
       "      <td>ENSG00000104529</td>\n",
       "      <td>0</td>\n",
       "      <td>blank|bridge</td>\n",
       "      <td>E0.U.122352|E1.L.11467</td>\n",
       "      <td>1298</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>144668897</td>\n",
       "      <td>144668899</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname strand        clv       aclv gene_name          gene_id  signed_dist_to_aclv evidence_type              contig_ids max_contig_len max_contig_mapq any_contig_is_hardclipped num_suffix_contigs num_bridge_contigs num_link_contigs num_blank_contigs num_total_contigs num_suffix_reads  \\\n",
       "781363  chr8    +      144677134  144682485  TIGD5     ENSG00000179886 -5351                 blank         E1.L.11467              1298           40              False                     0                  0                  0                1                 1                 0                 \n",
       "797962  chr8    -      144663470  144663248  EEF1D     ENSG00000104529  222                  blank         E0.S.110148|E1.L.11467  1298           40              False                     0                  0                  0                2                 2                 0                 \n",
       "797971  chr8    -      144668898  144668898  EEF1D     ENSG00000104529  0                    blank|bridge  E0.U.122352|E1.L.11467  1298           40              False                     0                  1                  0                1                 2                 0                 \n",
       "\n",
       "       max_suffix_contig_tail_len num_bridge_reads max_bridge_read_tail_len num_link_reads ctg_hex ctg_hex_id ctg_hex_pos ctg_hex_dist ref_hex ref_hex_id ref_hex_pos ref_hex_dist     clv_m1     clv_p1  \n",
       "781363  0                          0                0                        0              NA      -1         -1          -1           NA      -1         -1          -1           144677133  144677135  \n",
       "797962  0                          0                0                        0              NA      -1         -1          -1           NA      -1         -1          -1           144663469  144663471  \n",
       "797971  0                          1                1                        0              NA      -1         -1          -1           NA      -1         -1          -1           144668897  144668899  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cur_df[cur_df.contig_ids.str.contains('E1.L.11467')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>strand</th>\n",
       "      <th>clv</th>\n",
       "      <th>aclv</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>signed_dist_to_aclv</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>contig_ids</th>\n",
       "      <th>max_contig_len</th>\n",
       "      <th>max_contig_mapq</th>\n",
       "      <th>any_contig_is_hardclipped</th>\n",
       "      <th>num_suffix_contigs</th>\n",
       "      <th>num_bridge_contigs</th>\n",
       "      <th>num_link_contigs</th>\n",
       "      <th>num_blank_contigs</th>\n",
       "      <th>num_total_contigs</th>\n",
       "      <th>num_suffix_reads</th>\n",
       "      <th>max_suffix_contig_tail_len</th>\n",
       "      <th>num_bridge_reads</th>\n",
       "      <th>max_bridge_read_tail_len</th>\n",
       "      <th>num_link_reads</th>\n",
       "      <th>ctg_hex</th>\n",
       "      <th>ctg_hex_id</th>\n",
       "      <th>ctg_hex_pos</th>\n",
       "      <th>ctg_hex_dist</th>\n",
       "      <th>ref_hex</th>\n",
       "      <th>ref_hex_id</th>\n",
       "      <th>ref_hex_pos</th>\n",
       "      <th>ref_hex_dist</th>\n",
       "      <th>clv_m1</th>\n",
       "      <th>clv_p1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>657356</th>\n",
       "      <td>chr5</td>\n",
       "      <td>-</td>\n",
       "      <td>74017082</td>\n",
       "      <td>74017083</td>\n",
       "      <td>GFM2</td>\n",
       "      <td>ENSG00000164347</td>\n",
       "      <td>-1</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E0.L.64191|E0.U.323536</td>\n",
       "      <td>963</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>74017105</td>\n",
       "      <td>23</td>\n",
       "      <td>AATAAA</td>\n",
       "      <td>16</td>\n",
       "      <td>74017105</td>\n",
       "      <td>23</td>\n",
       "      <td>74017081</td>\n",
       "      <td>74017083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname strand       clv      aclv gene_name          gene_id  signed_dist_to_aclv evidence_type              contig_ids max_contig_len max_contig_mapq any_contig_is_hardclipped num_suffix_contigs num_bridge_contigs num_link_contigs num_blank_contigs num_total_contigs num_suffix_reads  \\\n",
       "657356  chr5    -      74017082  74017083  GFM2      ENSG00000164347 -1                    bridge        E0.L.64191|E0.U.323536  963            40              False                     0                  2                  0                0                 2                 0                 \n",
       "\n",
       "       max_suffix_contig_tail_len num_bridge_reads max_bridge_read_tail_len num_link_reads ctg_hex ctg_hex_id ctg_hex_pos ctg_hex_dist ref_hex ref_hex_id ref_hex_pos ref_hex_dist    clv_m1    clv_p1  \n",
       "657356  0                          4                5                        0              AATAAA  16         74017105    23           AATAAA  16         74017105    23           74017081  74017083  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cur_df.query('clv == 74017082')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# missed_df.query('evidence_type == \"bridge\"').query('signed_dist_to_aclv <= 0').sort_values('signed_dist_to_aclv', ascending=False)\\\n",
    "#     .head(10).reset_index(drop=True).iloc[5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>strand</th>\n",
       "      <th>clv</th>\n",
       "      <th>aclv</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>signed_dist_to_aclv</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>contig_ids</th>\n",
       "      <th>max_contig_len</th>\n",
       "      <th>max_contig_mapq</th>\n",
       "      <th>any_contig_is_hardclipped</th>\n",
       "      <th>num_suffix_contigs</th>\n",
       "      <th>num_bridge_contigs</th>\n",
       "      <th>num_link_contigs</th>\n",
       "      <th>num_blank_contigs</th>\n",
       "      <th>num_total_contigs</th>\n",
       "      <th>num_suffix_reads</th>\n",
       "      <th>max_suffix_contig_tail_len</th>\n",
       "      <th>num_bridge_reads</th>\n",
       "      <th>max_bridge_read_tail_len</th>\n",
       "      <th>num_link_reads</th>\n",
       "      <th>ctg_hex</th>\n",
       "      <th>ctg_hex_id</th>\n",
       "      <th>ctg_hex_pos</th>\n",
       "      <th>ctg_hex_dist</th>\n",
       "      <th>ref_hex</th>\n",
       "      <th>ref_hex_id</th>\n",
       "      <th>ref_hex_pos</th>\n",
       "      <th>ref_hex_dist</th>\n",
       "      <th>clv_m1</th>\n",
       "      <th>clv_p1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>812011</th>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>129638500</td>\n",
       "      <td>129648154</td>\n",
       "      <td>ZBTB34</td>\n",
       "      <td>ENSG00000177125</td>\n",
       "      <td>-9654</td>\n",
       "      <td>blank</td>\n",
       "      <td>01.U.576215</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>129638499</td>\n",
       "      <td>129638501</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname strand        clv       aclv gene_name          gene_id  signed_dist_to_aclv evidence_type   contig_ids max_contig_len max_contig_mapq any_contig_is_hardclipped num_suffix_contigs num_bridge_contigs num_link_contigs num_blank_contigs num_total_contigs num_suffix_reads  \\\n",
       "812011  chr9    +      129638500  129648154  ZBTB34    ENSG00000177125 -9654                 blank         01.U.576215  42             0               True                      0                  0                  0                1                 1                 0                 \n",
       "\n",
       "       max_suffix_contig_tail_len num_bridge_reads max_bridge_read_tail_len num_link_reads ctg_hex ctg_hex_id ctg_hex_pos ctg_hex_dist ref_hex ref_hex_id ref_hex_pos ref_hex_dist     clv_m1     clv_p1  \n",
       "812011  0                          0                0                        0              NA      -1         -1          -1           NA      -1         -1          -1           129638499  129638501  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cur_df.query('contig_ids == \"01.U.576215\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>strand</th>\n",
       "      <th>clv</th>\n",
       "      <th>aclv</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>signed_dist_to_aclv</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>contig_ids</th>\n",
       "      <th>max_contig_len</th>\n",
       "      <th>max_contig_mapq</th>\n",
       "      <th>any_contig_is_hardclipped</th>\n",
       "      <th>num_suffix_contigs</th>\n",
       "      <th>num_bridge_contigs</th>\n",
       "      <th>num_link_contigs</th>\n",
       "      <th>num_blank_contigs</th>\n",
       "      <th>num_total_contigs</th>\n",
       "      <th>num_suffix_reads</th>\n",
       "      <th>max_suffix_contig_tail_len</th>\n",
       "      <th>num_bridge_reads</th>\n",
       "      <th>max_bridge_read_tail_len</th>\n",
       "      <th>num_link_reads</th>\n",
       "      <th>ctg_hex</th>\n",
       "      <th>ctg_hex_id</th>\n",
       "      <th>ctg_hex_pos</th>\n",
       "      <th>ctg_hex_dist</th>\n",
       "      <th>ref_hex</th>\n",
       "      <th>ref_hex_id</th>\n",
       "      <th>ref_hex_pos</th>\n",
       "      <th>ref_hex_dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>288391</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>16388114</td>\n",
       "      <td>16388135</td>\n",
       "      <td>NOMO3</td>\n",
       "      <td>ENSG00000103226</td>\n",
       "      <td>-21</td>\n",
       "      <td>bridge</td>\n",
       "      <td>01.L.377168</td>\n",
       "      <td>184</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718349</th>\n",
       "      <td>chr7</td>\n",
       "      <td>+</td>\n",
       "      <td>5566820</td>\n",
       "      <td>5596877</td>\n",
       "      <td>CTB-161C1.1</td>\n",
       "      <td>ENSG00000272719</td>\n",
       "      <td>-30057</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E0.L.69552</td>\n",
       "      <td>267</td>\n",
       "      <td>40</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639318</th>\n",
       "      <td>chr5</td>\n",
       "      <td>+</td>\n",
       "      <td>117081763</td>\n",
       "      <td>116881994</td>\n",
       "      <td>LINC00992</td>\n",
       "      <td>ENSG00000248663</td>\n",
       "      <td>199769</td>\n",
       "      <td>bridge</td>\n",
       "      <td>01.U.677024</td>\n",
       "      <td>75</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706718</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>79918293</td>\n",
       "      <td>79910962</td>\n",
       "      <td>HMGN3</td>\n",
       "      <td>ENSG00000118418</td>\n",
       "      <td>7331</td>\n",
       "      <td>bridge</td>\n",
       "      <td>E2.L.1715</td>\n",
       "      <td>856</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629627</th>\n",
       "      <td>chr5</td>\n",
       "      <td>+</td>\n",
       "      <td>33631226</td>\n",
       "      <td>33503241</td>\n",
       "      <td>CTD-2233C11.3</td>\n",
       "      <td>ENSG00000249360</td>\n",
       "      <td>127985</td>\n",
       "      <td>bridge</td>\n",
       "      <td>01.U.1059438</td>\n",
       "      <td>75</td>\n",
       "      <td>40</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>NA</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname strand        clv       aclv      gene_name          gene_id  signed_dist_to_aclv evidence_type    contig_ids max_contig_len max_contig_mapq any_contig_is_hardclipped num_suffix_contigs num_bridge_contigs num_link_contigs num_blank_contigs num_total_contigs num_suffix_reads  \\\n",
       "288391  chr16   +      16388114   16388135   NOMO3          ENSG00000103226 -21                   bridge        01.L.377168   184            0               True                      0                  1                  0                0                 1                 0                 \n",
       "718349  chr7    +      5566820    5596877    CTB-161C1.1    ENSG00000272719 -30057                bridge        E0.L.69552    267            40              True                      0                  1                  0                0                 1                 0                 \n",
       "639318  chr5    +      117081763  116881994  LINC00992      ENSG00000248663  199769               bridge        01.U.677024   75             4               False                     0                  1                  0                0                 1                 0                 \n",
       "706718  chr6    -      79918293   79910962   HMGN3          ENSG00000118418  7331                 bridge        E2.L.1715     856            40              False                     0                  1                  0                0                 1                 0                 \n",
       "629627  chr5    +      33631226   33503241   CTD-2233C11.3  ENSG00000249360  127985               bridge        01.U.1059438  75             40              False                     0                  1                  0                0                 1                 0                 \n",
       "\n",
       "       max_suffix_contig_tail_len num_bridge_reads max_bridge_read_tail_len num_link_reads ctg_hex ctg_hex_id ctg_hex_pos ctg_hex_dist ref_hex ref_hex_id ref_hex_pos ref_hex_dist  \n",
       "288391  0                          67               2                        0              NA      -1         -1          -1           NA      -1         -1          -1           \n",
       "718349  0                          26               7                        0              NA      -1         -1          -1           NA      -1         -1          -1           \n",
       "639318  0                          15               24                       0              NA      -1         -1          -1           NA      -1         -1          -1           \n",
       "706718  0                          13               2                        0              NA      -1         -1          -1           NA      -1         -1          -1           \n",
       "629627  0                          12               10                       0              NA      -1         -1          -1           NA      -1         -1          -1           "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missed_df.query('evidence_type == \"bridge\"').sort_values('num_bridge_reads', ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
